enterprise/metrics/src/main/java/org/neo4j/metrics/source/JvmMetrics.java

```diff
@@ -0,0 +1,220 @@
+                    {
+                        try
+                        {
+                            emitter.removeNotificationListener( listener );
+                        }
+                        catch ( ListenerNotFoundException e )
+                        {
+                            throw new RuntimeException( e );
+                        }
+                    }
+                } );
+            }
+        }
+
+        // Memory metrics
+        if ( config.get( MetricsSettings.jvmMemoryEnabled ) )
+        {
+            for ( final MemoryPoolMXBean memPool : ManagementFactory.getMemoryPoolMXBeans() )
+            {
+                registry.register( name( MEMORY_POOL, prettifyName( memPool.getName() ) ), new Gauge<Long>()
+                {
+                    public Long getValue()
+                    {
+                        return memPool.getUsage().getUsed();
+                    }
+                } );
+            }
+        }
+
+        // Buffer pools
+        if ( config.get( MetricsSettings.jvmBuffersEnabled ) )
+        {
+            for ( final BufferPoolMXBean pool : ManagementFactory.getPlatformMXBeans( BufferPoolMXBean.class ) )
+            {
+                registry.register( name( MEMORY_BUFFER, prettifyName( pool.getName() ), "count" ), new Gauge<Long>()
+                {
+                    public Long getValue()
+                    {
+                        return pool.getCount();
+                    }
+                } );
+
+                registry.register( name( MEMORY_BUFFER, prettifyName( pool.getName() ), "used" ), new Gauge<Long>()
+                {
+                    public Long getValue()
+                    {
+                        return pool.getMemoryUsed();
+                    }
+                } );
+
+                registry.register( name( MEMORY_BUFFER, prettifyName( pool.getName() ), "capacity" ),
+                        new Gauge<Long>()
+                        {
+                            public Long getValue()
+                            {
+                                return pool.getTotalCapacity();
+                            }
+                        } );
+            }
+        }
+
+        // Threads
+        if ( config.get( MetricsSettings.jvmThreadsEnabled ) )
+        {
+            registry.register( name( THREAD, "count" ), new Gauge<Integer>()
+            {
+                public Integer getValue()
+                {
+                    return Thread.activeCount();
+                }
+            } );
+        }
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        // VM stats
+        // Garbage collection
+        if ( config.get( MetricsSettings.jvmGcEnabled ) )
+        {
+            for ( Runnable handle : removeListenerHandlers )
+            {
+                handle.run();
+            }
+            for ( GarbageCollectorMXBean gcBean : ManagementFactory.getGarbageCollectorMXBeans() )
```

Comment:
We should track the exact metric names that we register. The documentation says that GarbageCollectorMXBeans can be added and removed during runtime, so we are not guaranteed that this call will return the same list of beans, as the call we made in the constructor.